### PR TITLE
feat: Rust cargo lambda workflow

### PR DESCRIPTION
**Pull request by calavera**
This is a continuation of #174. It updates the Rust workflow to use [cargo-lambda](https://crates.io/crates/cargo-lambda), a tool that we've been building to simplify how to compile Rust functions for AWS Lambda. It supports cross-compiling natively, so you don't have to worry about it. The workflow is much more simple now, since it just calls cargo-lambda and copies the resulting artifacts.

I've also updated tests and example projects to work with this new version.
#174 can be closed, since this PR picks up where it left.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by calavera** [comment]
@jfuss can you help me figure out what's going on with Appveyor? I think at this point none of the errors I'm seeing are due to the changes in this PR.

**Comment by hawflau** [comment]
On the Linux Appveyor jobs, it seems all tests passed but the format check failed due to `click-8.1.2` breaking `black`. We can pin `click` to 8.0.4 as described in this [issue](https://github.com/psf/black/issues/2964)

I'm still investigating the Windows Appveyor jobs. Will update later.


```
Successfully installed appdirs-1.4.4 astroid-2.4.2 attrs-21.4.0 aws-lambda-builders-1.14.0 black-20.8b1 click-8.1.2 coverage-5.3 flake8-3.8.4 iniconfig-1.1.1 isort-5.10.1 lazy-object-proxy-1.4.3 mccabe-0.6.1 mock-4.0.2 mypy-extensions-0.4.3 packaging-21.3 parameterized-0.7.4 pathspec-0.9.0 pluggy-1.0.0 py-1.11.0 pycodestyle-2.6.0 pyelftools-0.28 pyflakes-2.2.0 pylint-2.6.2 pyparsing-3.0.7 pytest-7.1.1 pytest-cov-2.10.1 regex-2022.3.15 six-1.16.0 toml-0.10.2 tomli-2.0.1 typed-ast-1.5.2 typing-extensions-4.1.1 wrapt-1.14.0
```

```
if [[ $PYTHON_VERSION = "3"* ]]; then black --check setup.py tests aws_lambda_builders; fi
Traceback (most recent call last):
  File "/home/appveyor/venv3.9.10/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/home/appveyor/venv3.9.10/lib/python3.9/site-packages/black/__init__.py", line 6606, in patched_main
    patch_click()
  File "/home/appveyor/venv3.9.10/lib/python3.9/site-packages/black/__init__.py", line 6595, in patch_click
    from click import _unicodefun  # type: ignore
ImportError: cannot import name '_unicodefun' from 'click' (/home/appveyor/venv3.9.10/lib/python3.9/site-packages/click/__init__.py)
Command exited with code 1
Build failed
```

**Comment by mndeveci** [comment]
Windows checks have been failing with following error: 
```
"C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\libgimli-1e3e01ed4f561dc1.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\librustc_demangle-dd8dddcba7c13fee.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\libstd_detect-00b13d31c3420656.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\libhashbrown-e4c8460b81557bd5.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\librustc_std_workspace_alloc-600d46e5f0005455.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\libunwind-e4a08a2585b62d37.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\libcfg_if-fbc21ec567cb9dc7.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\liblibc-10b3fa8e49bd978f.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\liballoc-64625b73694ffce7.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\librustc_std_workspace_core-fb6b78ac543a58ee.rlib" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\libcore-1d1f2d1bec6f51b8.rlib" "-Wl,--end-group" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib\\libcompiler_builtins-acf5ff6e9595d982.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "C:\\Users\\appveyor\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-unknown-linux-gnu\\lib" "-o" "C:\\projects\\aws-lambda-builders\\tests\\integration\\workflows\\rust_cargo\\testdata\\workspaces\\target\\x86_64-unknown-linux-gnu\\debug\\deps\\foo-e440e43715e67040" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs"
E                 = note: warning: unsupported linker arg: -znoexecstack
E                         warning: unsupported linker arg: -zrelro
E                         warning: unsupported linker arg: -znow
E                         error(link): unable to spawn C:\Users\appveyor\zig-windows-x86_64-0.9.1\zig.exe: Unexpected
E                         error: UnableToSpawnSelf
E                         
E                         C:\projects\aws-lambda-builders\tests\integration\workflows\rust_cargo\testdata\workspaces>C:\Users\appveyor\.cargo\bin\cargo-lambda.exe zig cc -- -target x86_64-linux-gnu -g -Wl,--version-script=C:\Users\appveyor\AppData\Local\cargo-zigbuild\0.8.7\fcntl.map -include C:\Users\appveyor\AppData\Local\cargo-zigbuild\0.8.7\fcntl.h @C:\Users\appveyor\AppData\Local\Temp\1\rustc0vR5ZG\linker-arguments 
E                         
E               
E               error: could not compile `foo` due to previous error
aws_lambda_builders\workflow.py:309: WorkflowFailedError
=========================== short test summary info ===========================
FAILED tests/integration/workflows/rust_cargo/test_rust_cargo.py::TestRustCargo::test_builds_hello_project
FAILED tests/integration/workflows/rust_cargo/test_rust_cargo.py::TestRustCargo::test_builds_hello_project_for_arm64
FAILED tests/integration/workflows/rust_cargo/test_rust_cargo.py::TestRustCargo::test_builds_workspaces_project
============ 3 failed, 80 passed, 6 skipped in 1006.58s (0:16:46) =============
```

**Comment by calavera** [comment]
Yup, I'm aware of the error, I just don't know how to reproduce it unfortunately because the tests pass in my windows laptop. I have more information here if you're curious: https://github.com/cargo-lambda/cargo-lambda/issues/77

**Comment by mndeveci** [comment]
Can you check if you disable (or extend) this maximum path option for windows?
https://www.howtogeek.com/266621/how-to-make-windows-10-accept-file-paths-over-260-characters/

**Comment by calavera** [comment]
That option is disabled in my windows machine, I've never changed it:
image

**Comment by hawflau** [comment]
Converting the PR to draft as it seems it's still work in progress

**Comment by mndeveci** [comment]
@calavera do you have any updates regarding to this windows failures? I wonder if we can push this change under a feature flag, and noting that Windows might have issues that we are still investigating. 

Rust is getting traction from the community, and providing a native builder for SAM CLI would be good for our customers.

**Comment by mndeveci** [comment]
I saw windows ones succeeded in last try but they also took like couple of seconds. There is also a condition below in same file, you need to update there as well;
```
for:
  - matrix:
      only:
        - image: Visual Studio 2019
```

**Comment by mndeveci** [comment]
> I was finally able to reproduce this on my personal laptop. So at least I can investigate why it's happening.

That is great news! I will be preparing feature flag in SAM CLI so that we can use it with this builder.

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (mndeveci) [inline aws_lambda_builders/workflows/rust_cargo/actions.py:16]: I wonder if we should name these classes with `Rust` prefix? `RustBuilderError` for instance.

**Review by mndeveci — COMMENTED** [review]
Thanks for addressing previous comments @calavera !

In general it looks good for me. Can we also add some debug log messages so that we can undestand which `cargo` action is executed or which file is been copied or moved in the `RustBuildAction` and `RustCopyAndRenameAction` classes? In general, anything that can be useful when we need to investigate an issue while using this from SAM CLI.

Last thing before we release this out would be checking if certain beta feature flag is set, otherwise it should raise an error and stop building it. But that still requires us to create that feature flag in SAM CLI.

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (mndeveci) [inline aws_lambda_builders/workflows/rust_cargo/actions.py:115]: I think it would be nice to print `out` (and `err` if it is not empty) through logging. I saw you set debug environment variable, which is great to investigate what is going on with building here, but I assume we need to print process outputs as well so that user can see what is executed and what was the result there.

nit: and while doing that, it would be better if we can rename the `p` to something more verbose (`rust_process` or `cargo_process`)

**Review by mndeveci — COMMENTED** [review]
> @mndeveci I've added debug logs for the two actions. I also set `RUST_LOG=debug` in the environment when debug is enabled to print internal debug logs from cargo-lambda, that will give people extra information if they need to debug issues.

I think that is a great touch to get more details about the execution there. But we might need to print the process out or err (please see my comment below)

**Review by hawflau — APPROVED** [review]
@calavera Thanks for your contribution, and sorry for taking so long to review it. The PR looks good to me overall.

I notice the integration test case "workspaces" covers the use case of having multiple "sub-packages" within one package, and the Rust Cargo Workflow can specify which binary to build. As mentioned in DESIGN.md, it's intended to support the multiple Lambdas use case. I'm just starting to learn Rust, so I wonder if the following is also a valid setup for the multiple Lambdas use case:

Folder structure:
```
├── Cargo.lock
├── Cargo.toml
└── src
    ├── function_1.rs
    └── function_2.rs
```
with content of `Cargo.toml`:
```
[package]
# ...

[dependencies]
# ...

[[bin]]
name = "function1"
path = "src/function_1.rs"

[[bin]]
name = "function2"
path = "src/function_2.rs"
```

If this is also a valid setup, I think we can add a test case for this. But it's a blocker to merge this PR.

Follow-up items after we merge this PR:
1. Update SAM CLI to support using the Rust Cargo Lambda Workflow in `sam build` behind an experimental flag as @mndeveci suggested above. We will need a simple design to describe the change needed and reach agreement on UX. 
1. Update the [Rust init template](https://github.com/aws/aws-sam-cli-app-templates/tree/master/provided.al2/rust/cookiecutter-aws-sam-hello-rust) or create a new Rust init template as we will put the feature behind the experiment flag.
1. Update the [provided.al2 build image](https://github.com/aws/aws-sam-build-images/blob/develop/build-image-src/Dockerfile-provided_al2) to install Rust tool kits (or create a new build image?)

**Comment by hawflau** [comment]
> > If this is also a valid setup, I think we can add a test case for this. But it's a blocker to merge this PR.
> 
> It's already supported by setting the `handler` attribute. I can add a test here if you want to wait to merge it.

My bad for the typo. I mean, it's not a blocker to merge.

**Comment by hawflau** [comment]
> @hawflau I think I'm going to need a lot of guidance if you want me to do all those three things. Do you have any pointers to get examples from previous contributions?
> 
> > Update SAM CLI to support using the Rust Cargo Lambda Workflow in sam build behind an experimental flag as @mndeveci suggested above. We will need a simple design to describe the change needed and reach agreement on UX.
> > Update the [Rust init template](https://github.com/aws/aws-sam-cli-app-templates/tree/master/provided.al2/rust/cookiecutter-aws-sam-hello-rust) or create a new Rust init template as we will put the feature behind the experiment flag.
> > Update the [provided.al2 build image](https://github.com/aws/aws-sam-build-images/blob/develop/build-image-src/Dockerfile-provided_al2) to install Rust tool kits (or create a new build image?)

@calavera no, it's just a reminder for myself about the remaining action items, and discuss with Team about the prioritization of them.

However, as I synced up with @mndeveci, we need to also add the experimental flag in this PR. Basically, it's just a reverse of this [PR](https://github.com/aws/aws-lambda-builders/pull/383). If you don't mind, I can commit to this PR directly to add the flag.

**Comment by hawflau** [comment]
`experimentalCargoLambda` would be good. Thanks!

**Comment by mndeveci** [comment]
> 3. Update the [provided.al2 build image](https://github.com/aws/aws-sam-build-images/blob/develop/build-image-src/Dockerfile-provided_al2) to install Rust tool kits (or create a new build image?)

Given that rust (or cargo) can build for different os/arch types this step might be nice to have. What do you think @calavera ?

**Comment by mndeveci** [comment]
> @mndeveci I can help if you give me some guidance. I don't know how to provide that image.
> 
> Is the process to add a new image to https://github.com/aws/aws-sam-build-images/tree/develop/build-image-src ? If you're ok with adding more things to the provided.al2, I can do that too, whichever you prefer.

I think @hawflau meant this image particularly (given that Rust will probably running with `provided.al2` images) https://github.com/aws/aws-sam-build-images/blob/develop/build-image-src/Dockerfile-provided_al2

**Review by hawflau — CHANGES_REQUESTED** [review]
@calavera We've completed security review and it's good to move forward. Seems like we need to resolve some merge conflict before we can merge. Would you be able to do it? or I can do it as well.

**Comment by hawflau** [comment]
Since we've moved away from AppVeyor to GitHub Action for running tests, I think the change needed is to remove the appveyor files and update the GH Actions to run the rust cargo tests.

File: .github/workflows/build.yml
Comment (mndeveci) [inline .github/workflows/build.yml:310]: Why do we need this default here?

File: .github/workflows/build.yml
Comment (mndeveci) [inline .github/workflows/build.yml:350]: Will customer require to install this if they want to use this builder?

File: .github/workflows/build.yml
Comment (mndeveci) [reply .github/workflows/build.yml:310]: I see, can we remove and see how it goes? If it fails then we can add it back.

File: .github/workflows/build.yml
Comment (mndeveci) [reply .github/workflows/build.yml:350]: Do we have some guardrails in the implementation that will gracefully exit if this is not available in the customer's machine?

**Comment by hawflau** [comment]
@calavera There were a couple of linting issues, can you please fix them?
```
 aws_lambda_builders/workflows/rust_cargo/actions.py:141:121: E501 Line too long (124 > 120 characters)
aws_lambda_builders/workflows/rust_cargo/cargo_lambda.py:54:121: E501 Line too long (124 > 120 characters)
aws_lambda_builders/workflows/rust_cargo/cargo_lambda.py:55:121: E501 Line too long (136 > 120 characters)
```

There's also another [issue](https://github.com/aws/aws-lambda-builders/actions/runs/4106039927/jobs/7083755645) related to the recent `DEFAULT_BUILD_DIR` changes. We've recently added an option to allow the Function to be built from its source code dir rather than copying the source code to a temp dir and build from the temp dir. You can take a look at [here](https://github.com/aws/aws-lambda-builders/blob/4008bdd5ba39950f628b86aaece7ab7c91a7b794/aws_lambda_builders/workflows/go_modules/workflow.py#L18-L19) to see the 2 lines that need to be added into the Rust Workflow. I think we can try with `BuildDirectory.SOURCE` for `DEFAULT_BUILD_DIR` and `BuildInSourceSupport.EXCLUSIVELY_SUPPORTED` for `BUILD_IN_SOURCE_SUPPORT` since this is usually faster.

**Review by hawflau — CHANGES_REQUESTED** [review]

**Comment by mndeveci** [comment]
@calavera some linting issues in last changes;
```
aws_lambda_builders/utils.py:93:5: PLR0912 Too many branches (13/12)
aws_lambda_builders/workflow.py:71:9: PLR0912 Too many branches (13/12)
```

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (hawflau) [reply aws_lambda_builders/workflows/rust_cargo/actions.py]: We now use `ruff` instead of `pylint` for linting. I'll check how to fix them or disable the rule for now

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (mndeveci) [inline aws_lambda_builders/workflows/rust_cargo/actions.py:138]: Should this be a warning instead of debug?

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (mndeveci) [inline aws_lambda_builders/workflows/rust_cargo/actions.py:36]: nit: can you also update doc formats to numpy format (https://www.dataquest.io/blog/documenting-in-python-with-docstrings/)

We are not doing a good job to align on that in this library, but at least we are trying to address them with new additions.

File: aws_lambda_builders/workflows/rust_cargo/workflow.py
Comment (mndeveci) [inline aws_lambda_builders/workflows/rust_cargo/workflow.py:37]: nit: `get` would return `None` if key doesn't exist in the `dict`.
```suggestion
        handler = options.get("artifact_executable_name")
        flags = options.get("cargo_lambda_flags")
```

**Review by mndeveci — CHANGES_REQUESTED** [review]

**Review by hawflau — APPROVED** [review]

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (hawflau) [inline aws_lambda_builders/workflows/rust_cargo/actions.py:75]: Can we make this the default behavior and only remove this flag is `self._mode` is `BuildMode.DEBUG`? 
Reason is building without `--release` seems to produce a much larger artifact that exceeds Lambda's size limit. Defaulting it to build with `--release` seems to cause less deployment issue

File: aws_lambda_builders/workflows/rust_cargo/cargo_lambda.py
Comment (hawflau) [inline aws_lambda_builders/workflows/rust_cargo/cargo_lambda.py:94]: While this works, customer will see the process hangs for a while before the entire `cargo lambda` build log is printed. Can we stream the output instead?

We can replace L94-L102 with:
```
        stdout = ""
        # Create a buffer and use a thread to gather the stderr stream into the buffer
        stderr_buf = io.BytesIO()
        stderr_thread = threading.Thread(target=shutil.copyfileobj, args=(cargo_process.stderr, stderr_buf), daemon=True)
        stderr_thread.start()

        # Log every stdout line by iterating
        for line in cargo_process.stdout:
            decoded_line = line.decode("utf-8").strip()
            LOG.info(decoded_line)
            # Gather total stdout
            stdout += decoded_line

        # Wait for the process to exit and stderr thread to end.
        return_code = cargo_process.wait()
        stderr_thread.join()

        if return_code != 0:
            # Raise an Error with the appropriate value from the stderr buffer.
            raise CargoLambdaExecutionException(message=stderr_buf.getvalue().decode("utf8").strip())
        return stdout
```

**Review by hawflau — CHANGES_REQUESTED** [review]

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (hawflau) [reply aws_lambda_builders/workflows/rust_cargo/actions.py:75]: FYI, Go and Dotnet workflows also default to `BuildMode.RELEASE`:
- https://github.com/aws/aws-lambda-builders/blob/2ad17ffb6f4a81f82b458614f2cf7c992610f2e1/aws_lambda_builders/workflows/go_modules/builder.py#L50-L52
- https://github.com/aws/aws-lambda-builders/blob/2ad17ffb6f4a81f82b458614f2cf7c992610f2e1/aws_lambda_builders/workflows/dotnet_clipackage/actions.py#L87-L89

File: aws_lambda_builders/workflows/rust_cargo/actions.py
Comment (hawflau) [inline aws_lambda_builders/workflows/rust_cargo/actions.py:146]: I did some integration tests and found that this doesn't copy the permission bits of the executable. We should use [`copy2`](https://docs.python.org/3/library/shutil.html#shutil.copy2) instead.

**Review by hawflau — CHANGES_REQUESTED** [review]

**Review by mndeveci — APPROVED** [review]

**Review by hawflau — APPROVED** [review]

**Comment by milesgranger** [comment]
Awesome work @calavera (and to the reviewers!) :tada:

**Pull request merged**
